docs/docker.md

```diff
@@ -0,0 +1,162 @@
+# Contenedor Docker para pruebas
+
+En este documento se especifica el procedimiento para la elección y despliegue de un contenedor de docker para ejecución de tests unitarios sobre la aplicación.
+
+****
+
+### :page_with_curl: Criterios para elección de imagen base:
+
+Lo primero para desplegar un contenedor docker, es una imagen base sobre la que trabajaremos. Todas las imágenes consideradas deberán atender los siguientes criterios:
+
+1. ***Tamaño.*** Es el criterio más importante. Se debe elegir la imagen que tenga *el menor tamaño posible*. El motivo es que no debe contener nada más que lo imprescindible para la ejecución de tests. Como consecuencia inmediata de esto, tendremos entre otras cosas un despliegue rápido del contenedor (aunque notemos que esta operación sólo se realizará una única vez).
+1. ***Seguridad.*** En una imagen, existirán diferentes herramientas que vendrán incluidas por defecto. Se debe considerar una que disponga de la menor cantidad de vulnerabilidades importantes que puedan comprometer a la aplicación.
+
+Aunque habrá más cuestiones a contemplar (como se habla justo a continuación), la versión de la imagen seleccionada siempre será la última disponible `latest`, pues el propósito de este contenedor es que sea *para realizar pruebas*.
+
+
+
+#### :twisted_rightwards_arrows: Alternativas en la elección
+
+Con estos criterios principales, hay tres alternativas diferentes que podemos tener en cuenta para la elección de la imagen base.
+
+a) **Imágenes oficiales de Python:** Son imágenes publicadas que contienen la última versión de este lenguaje, y generalmente, gran parte de las bibliotecas del mismo.
+
+b) **Imágenes no oficiales de Python:** Son imágenes publicadas por algunas empresas externas, como  [Bitnami](https://hub.docker.com/u/bitnami) [CircleCI](https://hub.docker.com/u/circleci) u [Okteto](https://hub.docker.com/u/okteto), para lenguajes y servicios en GitHub.
+
+c) **Imágenes de Sistemas Operativos:** Son imágenes de sistemas operativos conocidos (Ubuntu, CentOS, etc...). Pueden considerarse como una alternativa a las otras dos, ya sean sus publicaciones oficiales, o alguna modificación de las mismas publicadas por otro usuario.
+
+
+
+Contemplando estas alternativas, hay que considerar criterios adicionales:
+
+*En el caso de elegir a) o b)*:
+
+* La **versión de Python** que debe contener es la última disponible (y estable). 
+* Disponer de la **utilidad `pip`**. Se debe verificar que dispone de pip, utilidad que permite instalar otras bibliotecas.
+
+
+
+*En el caso de tomar c):*
+
+* Disponer de **gestor de paquetes** u **otra utilidad de descarga/instalación** que permita instalar la última versión de Python, pues evidentemente, si no viene con la imagen habrá que instalarlo.
+* Disponer de la **utilidad `pip`**.
```

Comment:
¿Por qué? No hace falta para instalar poetry, y poe lo puedes instalar con poetry.

Reply:
Sé que poetry se puede instalar tanto con pip, como con `curl`. Pero realmente no me merece la pena instalar poetry con `curl`. En la imagen que he decidido usar no dispone de curl, y no me merece la pena instalarlo y luego desinstalarlo, sólo para este fin, cuando puedo instalar poetry y poe conjuntamente para el mismo propósito.

Y con respecto a poe y poetry, este asunto ya lo discutimos en el objetivo 3 (la interdependencia de estas dos herramientas). Había las dos posibilidades, una la que estoy considerando actualmente que es que poe no sea una dependencia intrínseca de poetry, y otra la que me comenta que sí lo sea, y en ese caso, cada vez que quisiera ejecutar una orden de poe tendría que hacer:

~~~bash
$ poetry run poe <orden>
~~~

Yo en mi proyecto indico que estas dos herramientas deben instalarse de manera global al usuario, me pareció la manera más fácil de uso para todo el mundo, tanto para mi, como para quien vea mi proyecto, así que considero que no me merece la pena incluir en el fichero `pyproject.toml` la dependencia de poe sólo para construirlo en el Dockerfile, y ese fue el criterio que decidí tomar para la elección de la imagen (a sabiendas que podría haber puesto también que tuviera curl por ejemplo, pero para qué lo iba a considerar si poe no se puede instalar de esa forma).

Reply:
> Sé que poetry se puede instalar tanto con pip, como con `curl`. Pero realmente no me merece la pena instalar poetry con `curl`. En la imagen que he decidido usar no dispone de curl, y no me merece la pena instalarlo y luego desinstalarlo, sólo para este fin, cuando puedo instalar poetry y poe conjuntamente para el mismo propósito.
> 

¿Puede que disponga de una herramienta similar a `curl`  para descargar cosas offline? Si no lo hiciera, ¿cómo descargaría sus propios paquetes?

En todo caso, date cuenta que estás poniendo como requisito algo que, o bien no es necesario, o bien simplemente se puede instalar en caso necesario. Eso te va a limitar mucho, por otro lado. La idea de elegir imagen es precisamente hacerlo con sólo lo estrictamente necesario.

> Y con respecto a poe y poetry, este asunto ya lo discutimos en el objetivo 3 (la interdependencia de estas dos herramientas). Había las dos posibilidades, una la que estoy considerando actualmente que es que poe no sea una dependencia intrínseca de poetry, y otra la que me comenta que sí lo sea, y en ese caso, cada vez que quisiera ejecutar una orden de poe tendría que hacer:
> 
> ```shell
> $ poetry run poe <orden>
> ```
> 

`poetry run` lo único que hace es buscar en los paths `bin` de las herramientas que ha instalado él. Si esos paths los estableces de forma independiente, puedes ejecutar `poe` donde te dé la gana. En general `poetry run` no hace nada mágico que no se puede hacer con un poco de conocimiento de Linux.

> Yo en mi proyecto indico que estas dos herramientas deben instalarse de manera global al usuario, me pareció la manera más fácil de uso para todo el mundo, tanto para mi, como para quien vea mi proyecto, así que considero que no me merece la pena incluir en el fichero `pyproject.toml` la dependencia de poe sólo para construirlo en el Dockerfile, y ese fue el criterio que decidí tomar para la elección de la imagen (a sabiendas que podría haber puesto también que tuviera curl por ejemplo, pero para qué lo iba a considerar si poe no se puede instalar de esa forma).

"De manera global" simplemente significa que estén accesibles en el PATH. Por eso uno de los elementos de la lista de comprobación es que os metáis para ver dónde y cómo está instalado todo. ¿Lo has hecho?